### PR TITLE
Feat/misc/notifications/nearby-users-only : Push notifications for new alerts only to nearby users

### PR DESCRIPTION
**Pull request by charliemangano**
# Push notifications for new alerts only to nearby users

## Description
This PR modifies the behavior of the alerts so that they're sent only to the users for which the alert is within their preferred distance. It closes issue #288.

## Changes
The main change in this PR is in the `push_new_alert` edge function (see PR #285 for explanations on edge functions). Instead of getting all the users from the database, this calls the server SQL function `get_tokens_of_users_within_alert_distance` (findable [here](https://supabase.com/dashboard/project/bhhjdcvdcfrxczbudraf/database/functions)). This makes it so that the computation remains server-side, which is good security-wise as we do not want the users' location exposed to everyone.

I needed each user to have a `preferred_distance`, so I made a new column in the `public.users` table in Supabase.

To compute each user's distance to the alert, I needed to store each user's location. For now, this is done through a new `locationGIS` column in the `public.users` table. We are aware that this causes security issues, since the location still ends up being exposed to the users, but we will be addressing it in another task (making a whole new table with only write access).

The user data classes (`User` and `UserDto`), the `LocationGIS` class and the related model (`UserModelSupabase`) had to be modified to accommodate these changes .

The location of each user is updated when a user lands on the app (`switchFromApproximateToPrecise`), when they quit it (in the background) (`switchFromPreciseToApproximate`) and when they close it (`cleanup`).

#### Added
None
#### Modified
- `supabase/functions/push_new_alert/index.ts`: updated `push_new_alert` to send notifications only to nearby users
- `app/src/main/java/com/android/periodpals/model/user/User.kt`, `app/src/main/java/com/android/periodpals/model/user/UserDto.kt`, `app/src/main/java/com/android/periodpals/model/location/LocationGIS.kt` and `app/src/main/java/com/android/periodpals/model/user/UserModelSupabase.kt`: modified to accommodate for new definitions of classes
- `app/src/main/java/com/android/periodpals/services/GPSServiceImpl.kt`: added uploading user's location logic
- `app/src/main/java/com/android/periodpals/MainActivity.kt`: adapted `GPSServiceImpl` initialization to call with user VM as argument 
- `app/src/test/java/com/android/periodpals/model/user/UserDtoTest.kt`, `app/src/test/java/com/android/periodpals/model/location/LocationGISDataClassTest.kt`, `app/src/test/java/com/android/periodpals/model/user/UserModelSupabaseTest.kt`, `app/src/test/java/com/android/periodpals/model/user/UserViewModelTest.kt`, `app/src/test/java/com/android/periodpals/services/GPSServiceImplTest.kt`, `supabase/tests/push_new_alert.test.ts`: added/modified tests for new funcitonalities

#### Removed
None

## Dependencies Added
None

## Testing
Tests were updated to accommodate for the new definitions of data classes as well as the new functionalities.

### CODE REVIEW

**Review by coaguila — APPROVED** [review]
## Summary
Great update with the new edge function and great idea to make it server-side to insure privacy and security for users. What made this review a bit hard was the many file effected but you couldn't do much about that due to the changes you had to do and another thing that made it hard was the lack of comments within the typescipt code which was hard to follow as a first time reader (although it was readable).
## Important (which includes)
#### Code Quality
LGTM
#### Functionality
LGTM
#### Performance
Performing the distance calculation should avoid putting to much on users devices.
#### Security
Going through the server to handle location based notifications is a great idea. We can be sure that user's location cant be found on other users devices unless transmitted by an alert.
## Testing
LGTM
## Documentation
Lacks a bit of comments for clarity but it is mostly a nitpick from my part.
## Suggestions
Add some more comments to your typescript code to make it more readable to others.

**Review by lazarinibruno — APPROVED** [review]
## Summary

Thanks for your contribution! It is really cool to see how you're slowly building the notifications to work as we originally intended 😼.

## Code quality

- Well-written and clear code.
- You added documentation and explanatory comments where needed.

## Tests

You adequately adapted the tests to the new implementation, and added new tests corresponding to the new features 🔥. Test pass locally.

## Comments

If I read the code correctly, you call `uploadUserLocation()` in the follwing cases:

- switchFromPreciseToApproximate
- switchFromApproximateToPrecise
- cleanup

Suppose that a new user opens the app for the first time. They will receive notifications for which location? Since neither of the cases above has happened, is there something as a GIS "default location"? Maybe @agonzalez-r has an idea?

I was thinking if there is a way to solve this, but there's no perfect solution.

- We could upload the user location in `startFusedLocationClient()`, but I'm not sure whether the location state is updated quickly enough to have the actual location.
- Doing it in the location callback would mean uploading the location to the database every two seconds, which is far from ideal.

I don't think that this issue terrible, but maybe we can have a look at it later if we have the time to do so.

## Steps before merge

Please consider my comments. Otherwise, LGTM.

File: app/src/test/java/com/android/periodpals/services/GPSServiceImplTest.kt [hunk 1]
```diff
@@ -331,6 +352,73 @@ class GPSServiceImplTest {
...
+    val mockLong = 16.0
+
+    `when`(userViewModel.user).thenReturn(mutableStateOf(userNoLocation))
+    `when`(userViewModel.loadUser(any(), any())).doAnswer {
+      val onSuccess = it.arguments[0] as () -> Unit
+      onSuccess()
+    }
+
+    // set the private _location value
+    val locationField = GPSServiceImpl::class.java.getDeclaredField("_location")
```
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/test/java/com/android/periodpals/services/GPSServiceImplTest.kt:368]: I also just discovered it, it's amazing!

File: app/src/test/java/com/android/periodpals/services/GPSServiceImplTest.kt [hunk 2]
```diff
@@ -331,6 +352,73 @@ class GPSServiceImplTest {
...
+            "test name",
+            "test url",
+            "test description",
+            "test dob",
+            "test fcm",
+            parseLocationGIS(Location(mockLat, mockLong, "test location")),
+        )
+    verify(userViewModel).saveUser(eq(userExpected), any(), any())
+  }
+
```
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/test/java/com/android/periodpals/services/GPSServiceImplTest.kt:421]: I originally wanted to, but I ran into some problems while testing: since `uploadUserLocation` is called after all the logic has taken place (we want to location to be updated to be precise), we also have to mock all the rest of the logic in the `switchFromApproximateToPrecise` function. This wasn't the case in the two other functions since `uploadUserLocation` is called before the rest of the logic in those.
As the goal was mainly to test that `uploadUserLocation` worked properly, I preferred to not lose any more time.

[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/test/java/com/android/periodpals/services/GPSServiceImplTest.kt:421]: I see! Well, the coverage is at 84% so I don't see any problem leaving it as is.

**Comment by lazarinibruno** [comment]
> Hi @lazarinibruno! Thanks for the review! I'll answer to your questions here!
> 
> > Suppose that a new user opens the app for the first time. They will receive notifications for which location?
> 
> The location is initialized to Bern, Switzerland at first. However, as soon as they leave the app (or close it), their location will be updated. Since the users are not expected to stay on the app for long periods of time, this is okay (plus the other options are really hard to implement).
> 
> > We could upload the user location in startFusedLocationClient(), but I'm not sure whether the location state is updated quickly enough to have the actual location.
> 
> You're right, the location is not updated quickly enough, so this didn't work when I tried :(
> 
> We can work towards a better solution in a later task.

Yes! Sounds perfect :)

File: app/src/main/java/com/android/periodpals/model/user/User.kt [hunk 3]
```diff
@@ -15,14 +19,16 @@ data class User(
     val description: String,
     val dob: String,
     val fcmToken: String? = null,
+    val locationGIS: LocationGIS = parseLocationGIS(Location.DEFAULT_LOCATION),
 ) {
-  inline fun asUserDto(): UserDto {
+  fun asUserDto(): UserDto {
```
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/main/java/com/android/periodpals/model/user/User.kt:30]: Adding the docs makes the CI fail (not the local though). I haven't been able to find the cause or fix it, so I'll be reverting to the last commit where it passed (without the docs).

**Comment by charliemangano** [comment]
As said in the comment to your review @lazarinibruno, adding the docs weirdly does not work. 
> Adding the docs makes the CI fail (not the local though). I haven't been able to find the cause or fix it, so I'll be reverting to [the last commit where it passed](https://github.com/PeriodPals/periodpals/pull/294/commits/7deb953388b86234f3194f2311363ebcd8fea22a) (without the docs).

File: app/src/main/java/com/android/periodpals/model/user/User.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/model/user/User.kt:24]: Mhh very weird. Okay, lets leave it like this then.
